transformer_engine/paddle/layer/base.py

```diff
@@ -0,0 +1,108 @@
+        _cublas_workspace = paddle.empty(
+            [get_cublas_workspace_size_bytes()],
+            dtype='uint8',
+        )
+    return _cublas_workspace
+
+
+class TransformerEngineBaseLayer(paddle.nn.Layer, ABC):
+    """Base TE Layer."""
+
+    def __init__(
+        self,
+        backend: str = 'transformer_engine',
+    ) -> None:
+        super().__init__()
+        assert 'gpu' in paddle.device.get_device(), "TransformerEngine needs CUDA."
+        self.backend = backend
+
+    def set_activation_dtype(self, inp: paddle.Tensor) -> None:
+        """Get activation data type for AMP."""
+        # Native AMP (`paddle.amp.auto_cast`) gets highest priority
+        tracer = _dygraph_tracer()
+        if tracer and tracer._amp_level != core.AmpLevel.O0:
+            if tracer._amp_dtype == 'float32':
+                self.activation_dtype = paddle.float32
+            elif tracer._amp_dtype == 'bfloat16':
+                self.activation_dtype = paddle.bfloat16
+            elif tracer._amp_dtype == 'float16':
+                self.activation_dtype = paddle.float16
+            else:
+                raise RuntimeError(f"AMP format {tracer._amp_dtype} is not supported.")
+            return
+
+        # All checks after this have already been performed once, thus skip
+        # We assume that user doesn't change input types across iterations
+        if hasattr(self, "activation_dtype") and self.activation_dtype == inp.dtype:
+            return
+
+        dtype = inp.dtype
+
+        for name, param in self.named_parameters():
+            if param is not None:
+                assert dtype == param.dtype, (
+                    "Data types for parameters must match when outside of autocasted region. "
+                    f" Found input dtype: {dtype} and {name!r} dtype: {param.dtype}")
+
+        self.activation_dtype = dtype
+
+    @contextmanager
+    def prepare_forward(
+        self,
+        inp: paddle.Tensor,
+    ) -> None:
+        """Checks and prep for FWD.
+        The context manager is needed because there isn't a way for a module to know
+        if it's the last FP8 module in the forward autocast. It is useful
+        to setup the forward aggregated amax reduction for every module
+        just in case. The autocast exit will pick up the most recent one.
+        """
+
+        self.set_activation_dtype(inp)
+
+        with nvtx_range(self.__class__.__name__ + " forward"):
+            yield inp
+
+    @abstractmethod
+    def _te_forward(self):
+        """TransformerEngine Path"""
+
+    @abstractmethod
+    def _pd_forward(self):
+        """Paddle Path"""
+
+    def forward(self, *args, **kwargs):
+        """forward"""
+        if self.backend == 'transformer_engine':
+            return self._te_forward(*args, **kwargs)
+        if self.backend == 'paddle':
+            return self._pd_forward(*args, **kwargs)
+        raise AttributeError(f"Backend {self.backend} is not supported.")
```

Comment:
What is the expected use case for the pure Paddle backend, and why use Transformer Engine in that case? If it is a reference implementation for testing, it would be better to move this logic to the tests. The pure Paddle backend won't support FP8 or model parallelism, and it adds an extra implementation burden when adding new modules.

Reply:
+1, also these names aren't the most accurate or easy to understand. We are in the TE library and having backend as `transformer_engine` vs `paddle` can be confusing

Reply:
The intent is to allow user to effortlessly switch to Paddle ops, which provides easy performance comparison (Paddle <-> TE BF16 or Paddle <-> TE FP8). It will be much easier to collect per-layer benchmark and spot performance 
issues (e.g., when TE is slower than expected). Without Paddle backend, when you want to do such benchmark, you have to rewrite the whole model, replacing all TE layers with Paddle layers and manually align the per-layer benchmarks.

@jeng1220 do you have further comments?

Reply:
@timmoon10 , @ksivaman ,
1. Baidu is also improving their BF16 kernels, and planning native FP8 support of Paddle. The purpose of this design is to make performance comparison easier, and help us to capture performance issues, and isolate performance issues. The comparison scope can be single layer, single block or even a model. And adding `_pd_forward` doesn't take much effort, only a few lines of code.

2. Further, Paddle has deep learning compiler too. It is expected to achieve automatic kernel fusion in the near future. As we know, there are many LLM variants, which have many different fusion patterns, such as:
    * layernorm_forward -> cast_transpose
    * layernorm_backward -> cast_transpose
    * layernorm_1_plus_forward -> cast_transpose
    * layernorm_1_plus_backward -> cast_transpose
    * rmsnorm_forward -> cast_transpose
    * rmsnorm_backward -> cast_transpose
    * gelu -> cast_transpose
    * dgelu -> cast_transpose
    * gated_gelu -> cast_transpose
    * dgated_gelu -> cast_transpose
    * ... various patterns
    It will be great if the compiler is able to generate fast kernel and achieve SOL for those patterns.
    But the compiler only recognizes Paddle Primitives instead of the custom kernels of Transformer Engine.
    Thus, I think we need to keep both `_pd_forward`  and  `_te_forward`  path.

If the backend naming is an issue, we can think another better name, such as `custom-call` vs `paddle`?

Reply:
I think the backend names of `paddle` and `transformer_engine` are fine since they describe who is responsible for launching the CUDA kernels. The confusing names reflect the design.

I see the value in switching between different backends. I suppose it is similar to the PyTorch attention logic that switches between plain PyTorch, cuDNN, and FlashAttention. But I don't think this logic belongs in the base class and it should be moved into the individual classes. This way we don't force each layer to have exactly one Transformer Engine impl and exactly one Paddle impl. Consider layers that are purely Paddle, or have multiple Paddle impls, or use other libraries (e.g. FlashAttention).

Reply:
Done.